### PR TITLE
refactor(sessiond): move events and update report callback into LocalEnforcer from LocalSessionManagerHandler

### PR DESCRIPTION
**Pull request by veshkemburu**
Signed-off-by: veshkemburu <veshkemburu@fb.com>

## Summary

After receiving stats from pipelined, sessiond performs actions on them such as reporting and record aggregation

## Test Plan

Transfer internal function callback to separate function. 
Move each LocalSessionManagerHandler function to LocalEnforcer
Add calls from LocalSessionManagerHandler to enforcer using local enforcer object

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Review by themarwhal — CHANGES_REQUESTED** [review]
Please add a summary to indicate why this change is necessary. Thanks :)

File: lte/gateway/c/session_manager/LocalEnforcer.cpp [hunk 1]
```diff
@@ -147,6 +148,123 @@ void LocalEnforcer::setup(
...
+        }
+        // Success!
+        update_session_credits_and_rules(
+            *session_map_ptr, response, session_uc);
+        report_session_update_event(*session_map_ptr, updates_by_session);
+        session_store_.update_sessions(session_uc);
+      });
+}
+
+void LocalEnforcer::HandlePipelinedResponse(
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/LocalEnforcer.cpp]: `HandlePipelinedResponse` -> `handle_pipelined_response`

File: lte/gateway/python/magma/pipelined/app/enforcement_stats.py [hunk 2]
```diff
@@ -577,6 +577,8 @@ def get_stats(self, cookie: int = 0, cookie_mask: int = 0):
         Use Ryu API to send a stats request containing cookie and cookie mask, retrieve a response and 
         convert to a Rule Record Table
         """
+        if not self._datapath:
```
Comment (themarwhal) [inline lte/gateway/python/magma/pipelined/app/enforcement_stats.py]: Should this be a separate PR?

File: lte/gateway/c/session_manager/LocalEnforcer.cpp [hunk 3]
```diff
@@ -147,6 +148,123 @@ void LocalEnforcer::setup(
...
+    MLOG(MDEBUG) << "Aggregating " << resp.records_size() << " records";
+    aggregate_records(session_map, resp, update);
+
+    check_usage_for_reporting(std::move(session_map), update);
+  }
+}
+
+void LocalEnforcer::poll_stats_enforcer() {
+  pipelined_client_->poll_stats(
+      0, 0, std::bind(&LocalEnforcer::HandlePipelinedResponse, this, _1, _2));
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/LocalEnforcer.cpp]: Can you add a comment on why we need to bind the arguments here?

File: lte/gateway/c/session_manager/LocalEnforcer.cpp [hunk 4]
```diff
@@ -147,6 +148,123 @@ void LocalEnforcer::setup(
...
+  session_store_.set_and_save_reporting_flag(true, request, session_uc);
+
+  // Before reporting and returning control to the event loop, increment the
+  // request numbers stored for the sessions in SessionStore
+  session_store_.sync_request_numbers(session_uc);
+
+  // report to cloud
+  reporter_->report_updates(
+      request,
+      [this, request, session_uc,
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/LocalEnforcer.cpp]: If this callback function is shared between here and LocalsessionManagerHandler, pull the callback function into a single function defined in `SessionReporter`. 
Always try to minimize the amount of code you have to copy :)

File: lte/gateway/python/magma/pipelined/app/enforcement_stats.py (hunk 2 above)
Comment (veshkemburu) [reply lte/gateway/python/magma/pipelined/app/enforcement_stats.py]: It's a minor adjustment to handle an edge case, I can make it a separate PR if desired but I thought it would be easier to just make the fix and put it in with this PR.

Comment (themarwhal) [reply lte/gateway/python/magma/pipelined/app/enforcement_stats.py]: Let's separate it out :) 
Each PR should handle independent changes
Also, I think if we separate it out, we can get it into master faster as it only requires @koolzz 's approval.

File: lte/gateway/c/session_manager/LocalEnforcer.cpp [hunk 5]
```diff
@@ -147,13 +148,127 @@ void LocalEnforcer::setup(
   }
 }
 
-void LocalEnforcer::poll_stats_enforcer() {
-  pipelined_client_->poll_stats(0, 0, [](Status status, RuleRecordTable resp) {
-    if (!status.ok()) {
-      MLOG(MERROR) << "Could not successfully poll stats: "
-                   << status.error_message();
+void LocalEnforcer::report_session_update_event(
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/LocalEnforcer.cpp:151]: Should the corresponding function in LocalSessionManagerHandler removed in this PR?

File: lte/gateway/c/session_manager/LocalEnforcer.cpp [hunk 6]
```diff
@@ -147,13 +148,127 @@ void LocalEnforcer::setup(
...
+                     << " because it couldn't be found";
+      continue;
     }
-  });
+    events_reporter_->session_updated(
+        session_id, (**session_it)->get_config(), it.second);
+  }
+}
+
+void LocalEnforcer::report_session_update_event_failure(
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/LocalEnforcer.cpp]: Same deal here

Comment (themarwhal) [reply lte/gateway/c/session_manager/LocalEnforcer.cpp]: Also, while you're at it can you rename this function to `report_session_update_failure_event`?

File: lte/gateway/c/session_manager/LocalEnforcer.cpp [hunk 7]
```diff
@@ -147,13 +148,127 @@ void LocalEnforcer::setup(
...
+    MLOG(MERROR) << "Could not successfully poll stats: "
+                 << status.error_message();
+  } else {
+    auto session_map = session_store_.read_all_sessions();
+    SessionUpdate update =
+        SessionStore::get_default_session_update(session_map);
+    MLOG(MDEBUG) << "Aggregating " << resp.records_size() << " records";
+    aggregate_records(session_map, resp, update);
+
+    check_usage_for_reporting(std::move(session_map), update);
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/LocalEnforcer.cpp]: Instead of `std::move` ing the entire map, you can pass by reference here.

```
void LocalEnforcer::check_usage_for_reporting(
     SessionMap& session_map, SessionUpdate& updates)
```

File: lte/gateway/c/session_manager/LocalEnforcer.h [hunk 8]
```diff
@@ -294,6 +294,27 @@ class LocalEnforcer {
...
+
+  /*
+   * Report flow stats from pipelined and track the usage per rule
+   */
+  void check_usage_for_reporting(
+      SessionMap session_map, SessionUpdate& session_uc);
+
+  void handle_pipelined_response(Status status, RuleRecordTable resp);
+
+  void report_updates_callback(
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/LocalEnforcer.h]: `report_updates_callback` -> `handle_session_update_response`

File: lte/gateway/c/session_manager/LocalEnforcer.h [hunk 9]
```diff
@@ -294,6 +294,27 @@ class LocalEnforcer {
...
+      UpdateSessionRequest& request,
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/LocalEnforcer.h]: Should this be const as the linter suggests?

**Review by themarwhal — CHANGES_REQUESTED** [review]

**Comment by themarwhal** [comment]
please add a test plan + fix PR title (`refactor(sessiond): move events and update report callback into LocalEnforcer`) .

File: lte/gateway/c/session_manager/LocalEnforcer.cpp (hunk 7 above)
Comment (veshkemburu) [reply lte/gateway/c/session_manager/LocalEnforcer.cpp]: Tried this before but it errors out.

**Review by themarwhal — APPROVED** [review]
